# Домашнее задание на проверку знаний библиотеки для анализа данных Pandas.

Вам представлен датасет, который содержит информацию по результатам сезона 2016 года игры в бейсбол.

Вам необходимо загрузить датасет (он находится в архиве с текущим домашним заданием). Провести его очистку, если это необходимо и ответить на поставленные аналитические
вопросы.

Список колонок, представленных в датасете:

---
* attendance        - количество зрителей матча
* away_team         - название гостевой команды
* away_team_errors  - количество ошибок гостевой команды
* away_team_hits    - количество хитов гостевой команды (хит в бейсболе - удар при котором отбивающий достиг первой базы)
* away_team_runs    - количество ранов гостевой команды (ран в бейсболе - очко заработанное игроком нападения)
* date              - дата проведения матча
* field_type        - тип игрового поля
* game_type         - тип игры
* home_team         - название домашней команды
* home_team_errors  - количество ошибок домашней команды
* home_team_hits    - количество хитов домашней команды (хит в бейсболе - удар при котором отбивающий достиг первой базы)
* home_team_runs    - количество ранов домашней команды (ран в бейсболе - очко заработанное игроком нападения)
* start_time        - время начала игрового матча
* venue             - название места проведения матча (стадион, поле, арена)
* day_of_week       - день недели проведения матча
* temperature       - температура воздуха в день проведения матча в фаренгейтах
* wind_speed        - скорость ветра в день проведения матча
* wind_direction    - направление скорости ветра
* sky               - облачность
* total_runs        - общее количество ранов для двух команд
* game_hours_dec    - длительность матча указаная в часах
* season            - тип игрового сезона
* home_team_win     - результат домашней команды (1 - выигрыш)
* home_team_loss    - результат домашней команды (0 - проигрыш)
* home_team_outcome - исход матча
---

Всего в задании 20 вопросов. За каждый правильный вопрос начисляется 5 баллов. Таким образом за все правильные ответы вы получите - 100 баллов. Оценка затем будет приведена к 10-ти бальной шкале.

---

**Кроме того, баллы могут быть снижены за**:

- 1 балл снижается за качество кода - старайтесь чтобы код был максимально лаконичным. Названия переменных (если они используются) должны отражать максимальную ее сущность. Вывод старайтесь делать минимальным, так как первичная цель - сделать качественный анализ, чтобы быстро ответить на поставленный вопрос (никаких таблиц на 100 строк).


В бейсболе есть несколько ключевых терминов, которые обозначают разные аспекты игры:

Матч (игра): Это отдельное соревнование между двумя командами. Каждый матч состоит из девяти иннингов, в течение которых команды поочередно атакуют и защищаются. Цель каждой команды — набрать больше очков (ранов), чем соперн
ик.
Сезон: Это период времени, в течение которого проводятся матчи. В профессиональном бейсболе сезон обычно длится несколько месяцев и включает в себя множество матчей между различными командами. Сезон может быть регулярным, в котором команды соревнуются за место в плей-офф, и постсезонным, где команды с лучшими результатами борются за чемпионс
тво.
Таким образом, матч — это единичное событие, а сезон — это совокупность матчей, которые проводятся в течение определенного времени.

In [3]:
# Импортируйте все необходимые библиотеки и модули которые вам необходимы
import numpy as np
import pandas as pd
import scipy.stats as st

In [4]:
# Загрузите датасет и проведите его очистку и форматирование если это необходимо
data = pd.read_csv('../practical_task1/baseball_games.csv', index_col=0)
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2463 entries, 0 to 2462
Data columns (total 25 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   attendance         2460 non-null   float64
 1   away_team          2463 non-null   object 
 2   away_team_errors   2463 non-null   int64  
 3   away_team_hits     2463 non-null   int64  
 4   away_team_runs     2463 non-null   int64  
 5   date               2463 non-null   object 
 6   field_type         2463 non-null   object 
 7   game_type          2463 non-null   object 
 8   home_team          2463 non-null   object 
 9   home_team_errors   2463 non-null   int64  
 10  home_team_hits     2463 non-null   int64  
 11  home_team_runs     2463 non-null   int64  
 12  start_time         2463 non-null   object 
 13  venue              2463 non-null   object 
 14  day_of_week        2463 non-null   object 
 15  temperature        2463 non-null   float64
 16  wind_speed         2463 non-n

In [5]:
data.head()

attendance              away_team  away_team_errors  away_team_hits  \
0     40030.0          New York Mets                 1               7   
1     21621.0  Philadelphia Phillies                 0               5   
2     12622.0        Minnesota Twins                 0               5   
3     18531.0   Washington Nationals                 0               8   
4     18572.0       Colorado Rockies                 1               8   

   away_team_runs        date field_type   game_type             home_team  \
0               3  2016-04-03   on grass  Night Game    Kansas City Royals   
1               2  2016-04-06   on grass  Night Game       Cincinnati Reds   
2               2  2016-04-06   on grass  Night Game     Baltimore Orioles   
3               3  2016-04-06   on grass  Night Game        Atlanta Braves   
4               4  2016-04-06   on grass    Day Game  Arizona Diamondbacks   

   home_team_errors  ...  temperature  wind_speed         wind_direction  \
0                 0  ...         74.0        14.0     from Right to Left   
1                 0  ...         55.0        24.0     from Right to Left   
2                 0  ...         48.0         7.0       out to Leftfield   
3                 1  ...         65.0        10.0     from Right to Left   
4                 0  ...         77.0         0.0   in unknown direction   

        sky total_runs  game_hours_dec          season home_team_win  \
0     Sunny          7        3.216667  regular season             1   
1  Overcast          5        2.383333  regular season             1   
2   Unknown          6        3.183333  regular season             1   
3    Cloudy          4        2.883333  regular season             0   
4   In Dome          7        2.650000  regular season             0   

  home_team_loss  home_team_outcome  
0              0                Win  
1              0                Win  
2              0                Win  
3              1               Loss  
4              1               Loss  

[5 rows x 25 columns]

In [6]:
data.describe().T

count          mean          std      min           25%  \
attendance        2460.0  30370.703659  9875.466703  8766.00  22432.000000   
away_team_errors  2463.0      0.580593     0.793391     0.00      0.000000   
away_team_hits    2463.0      8.764515     3.511581     1.00      6.000000   
away_team_runs    2463.0      4.413723     3.104556     0.00      2.000000   
home_team_errors  2463.0      0.585871     0.805542     0.00      0.000000   
home_team_hits    2463.0      8.611855     3.436965     0.00      6.000000   
home_team_runs    2463.0      4.519691     3.111572     0.00      2.000000   
temperature       2463.0     73.818920    10.583106    31.00     67.000000   
wind_speed        2463.0      7.494925     5.055970     0.00      4.000000   
total_runs        2463.0      8.933415     4.480352     1.00      6.000000   
game_hours_dec    2463.0      3.085302     0.460491     1.25      2.783333   
home_team_win     2463.0      0.529436     0.499234     0.00      0.000000   
home_team_loss    2463.0      0.470158     0.499210     0.00      0.000000   

                           50%           75%           max  
attendance        30604.500000  38396.250000  54449.000000  
away_team_errors      0.000000      1.000000      5.000000  
away_team_hits        8.000000     11.000000     22.000000  
away_team_runs        4.000000      6.000000     21.000000  
home_team_errors      0.000000      1.000000      5.000000  
home_team_hits        8.000000     11.000000     22.000000  
home_team_runs        4.000000      6.000000     17.000000  
temperature          74.000000     81.000000    101.000000  
wind_speed            7.000000     11.000000     25.000000  
total_runs            8.000000     12.000000     29.000000  
game_hours_dec        3.033333      3.308333      6.216667  
home_team_win         1.000000      1.000000      1.000000  
home_team_loss        0.000000      1.000000      1.000000

Обратим внимание на пропущенные значения в столбце 'attendance' (количество зрителей матча) 
заменим пустые значения столбца attendance (количество зрителей матча) на медиану 
(т.к. с пустыми значениями в данных лучше не работать)

In [8]:
median_attendance = data['attendance'].median()
data['attendance'] = data['attendance'].fillna(median_attendance)

data - наш исходный датасет, сохраним отдельно data_rs - копия data, но исключительно с типом игрового сезона == 'regular season'

In [10]:
data_rs = data[data['season'] == 'regular season']
data_rs.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2428 entries, 0 to 2462
Data columns (total 25 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   attendance         2428 non-null   float64
 1   away_team          2428 non-null   object 
 2   away_team_errors   2428 non-null   int64  
 3   away_team_hits     2428 non-null   int64  
 4   away_team_runs     2428 non-null   int64  
 5   date               2428 non-null   object 
 6   field_type         2428 non-null   object 
 7   game_type          2428 non-null   object 
 8   home_team          2428 non-null   object 
 9   home_team_errors   2428 non-null   int64  
 10  home_team_hits     2428 non-null   int64  
 11  home_team_runs     2428 non-null   int64  
 12  start_time         2428 non-null   object 
 13  venue              2428 non-null   object 
 14  day_of_week        2428 non-null   object 
 15  temperature        2428 non-null   float64
 16  wind_speed         2428 non-n

# 1. На какую игру пришло максимальное количество зрителей за весь сезон игр?

столбец 'attendance' показывает количество зрителей, соответственно необходимо взять максимальное значение этого столбца из датасета с фильтром по регулярному сезону

In [13]:
print("Матчи с максимальным количеством зрителей за весь сезон (тип игрового сезона == 'regular season'):")
data_rs[data_rs['attendance'] == max(data_rs['attendance'])][['home_team', 'away_team', 'venue', 'date', 'attendance']]

Матчи с максимальным количеством зрителей за весь сезон (тип игрового сезона == 'regular season'):


home_team             away_team           venue        date  \
534  Los Angeles Dodgers  San Francisco Giants  Dodger Stadium  2016-09-20   

     attendance  
534     53621.0

# 2. Какая игра была самая холодная (temperature) за весь сезон?

столбец 'temperature' показывает  температуру воздуха в день проведения матча в фаренгейтах, соответственно необходимо взять минимальное значение этого столбца из датасета с фильтром по регулярному сезону

In [16]:
print("Матчи с самой холодной температурой за весь сезон (тип игрового сезона == 'regular season'):")
data_rs[data_rs['temperature'] == min(data_rs['temperature'])][['home_team', 'away_team', 'venue', 'date', 'temperature']]

Матчи с самой холодной температурой за весь сезон (тип игрового сезона == 'regular season'):


home_team         away_team          venue        date  temperature
2409  Detroit Tigers  New York Yankees  Comerica Park  2016-04-09         31.0

# 3. Какая игра была самая теплая за весь сезон?

столбец 'temperature' показывает  температура воздуха в день проведения матча в фаренгейтах, соответственно необходимо взять максимальное значение этого столбца из датасета с фильтром по регулярному сезону

In [19]:
print("Матчи с самой теплой температурой за весь сезон (тип игрового сезона == 'regular season'):")
data_rs[data_rs['temperature'] == max(data_rs['temperature'])][['home_team', 'away_team', 'venue', 'date', 'temperature']]

Матчи с самой теплой температурой за весь сезон (тип игрового сезона == 'regular season'):


home_team             away_team        venue        date  \
2026  Arizona Diamondbacks  San Francisco Giants  Chase Field  2016-05-13   

      temperature  
2026        101.0

# 4. Какая игра в сезоне была самая долгая по продолжительности матча?

столбец 'game_hours_dec' показывает  длительность матча, указанную в часах, соответственно необходимо взять максимальное значение этого столбца из датасета с фильтром по регулярному сезону

In [22]:
print("Матчи с самой долгой продолжительностью за весь сезон (тип игрового сезона == 'regular season'):")
data_rs[data_rs['game_hours_dec'] == max(data_rs['game_hours_dec'])][['home_team', 'away_team', 'venue', 'date', 'game_hours_dec']]

Матчи с самой долгой продолжительностью за весь сезон (тип игрового сезона == 'regular season'):


home_team          away_team          venue        date  \
1445  Toronto Blue Jays  Cleveland Indians  Rogers Centre  2016-07-01   

      game_hours_dec  
1445        6.216667

# 5. Какая игра в сезоне была самая короткая по продолжительности матча?

столбец 'game_hours_dec' показывает длительность матча указанную в часах, соответственно необходимо взять минимальное значение этого столбца из датасета с фильтром по регулярному сезону

In [25]:
print("Матчи с самой короткой продолжительностью за весь сезон (тип игрового сезона == 'regular season'):")
data_rs[data_rs['game_hours_dec'] == min(data_rs['game_hours_dec'])][['home_team', 'away_team', 'venue', 'date', 'game_hours_dec']]

Матчи с самой короткой продолжительностью за весь сезон (тип игрового сезона == 'regular season'):


home_team     away_team     venue        date  game_hours_dec
423  Pittsburgh Pirates  Chicago Cubs  PNC Park  2016-09-29            1.25

# 6. Сколько матчей в сезоне закончилось ничьей?

столбец 'home_team_win' показывает 1 в случае победы домашней команды, а столбец 'home_team_loss' показывает 1 в случае победы гостевой команды, т.о. чтобы определить матчи, в которых была ничья необходимо чтобы обе команды одновременно не выиграли или не проиграли (из датасета с фильтром по регулярному сезону)

In [28]:
print("Количество матчей в сезоне (тип игрового сезона == 'regular season'), закончившихся ничьей:")
data_rs[data_rs['home_team_win'] == data_rs['home_team_loss']].shape[0]

Количество матчей в сезоне (тип игрового сезона == 'regular season'), закончившихся ничьей:


1

# 7. Какая игра была последней в сезоне?

столбец 'date' показывает дату проведения матча, соответственно необходимо взять максимальное значение этого столбца из датасета с фильтром по регулярному сезону

In [31]:
print("Последняя игра в сезоне (тип игрового сезона == 'regular season'):")
data_rs[data_rs['date'] == max(data_rs['date'])][['home_team', 'away_team', 'venue', 'date']]

Последняя игра в сезоне (тип игрового сезона == 'regular season'):


home_team            away_team  \
349           Washington Nationals        Miami Marlins   
385                  Texas Rangers       Tampa Bay Rays   
386            St. Louis Cardinals   Pittsburgh Pirates   
387           San Francisco Giants  Los Angeles Dodgers   
388               Seattle Mariners    Oakland Athletics   
389          Philadelphia Phillies        New York Mets   
390               New York Yankees    Baltimore Orioles   
391             Kansas City Royals    Cleveland Indians   
392               Colorado Rockies    Milwaukee Brewers   
393                Cincinnati Reds         Chicago Cubs   
394              Chicago White Sox      Minnesota Twins   
395                 Boston Red Sox    Toronto Blue Jays   
396                 Atlanta Braves       Detroit Tigers   
397           Arizona Diamondbacks     San Diego Padres   
398  Los Angeles Angels of Anaheim       Houston Astros   

                            venue        date  
349                Nationals Park  2016-10-02  
385  Globe Life Park in Arlington  2016-10-02  
386             Busch Stadium III  2016-10-02  
387                     AT&T Park  2016-10-02  
388                  Safeco Field  2016-10-02  
389            Citizens Bank Park  2016-10-02  
390            Yankee Stadium III  2016-10-02  
391              Kauffman Stadium  2016-10-02  
392                   Coors Field  2016-10-02  
393      Great American Ball Park  2016-10-02  
394           U.S. Cellular Field  2016-10-02  
395                   Fenway Park  2016-10-02  
396                  Turner Field  2016-10-02  
397                   Chase Field  2016-10-02  
398      Angel Stadium of Anaheim  2016-10-02

# 8. У какой игры было минимальное количество зрителей?

столбец 'attendance' показывает количество зрителей, соответственно необходимо взять минимальное значение этого столбца из датасета (первоначального, без фильтра по сезону)

In [34]:
print("Матчи с минимальным количеством зрителей (не важно какой сезон):")
data[data['attendance'] == min(data['attendance'])][['home_team', 'away_team', 'venue', 'date', 'attendance']]

Матчи с минимальным количеством зрителей (не важно какой сезон):


home_team       away_team              venue        date  \
2130  Cleveland Indians  Detroit Tigers  Progressive Field  2016-05-04   

      attendance  
2130      8766.0

# 9. Какая игра в сезоне была самая ветренная?

столбец 'wind_speed' показывает скорость ветра в день проведения матча, соответственно необходимо взять максимальное значение этого столбца из датасета с фильтром по регулярному сезону

In [37]:
print("Матчи с максимальной скоростью ветра за весь сезон (тип игрового сезона == 'regular season'):")
data_rs[data_rs['wind_speed'] == max(data_rs['wind_speed'])][['home_team', 'away_team', 'venue', 'date', 'wind_speed']]

Матчи с максимальной скоростью ветра за весь сезон (тип игрового сезона == 'regular season'):


home_team          away_team        venue        date  \
1655  San Francisco Giants  Milwaukee Brewers    AT&T Park  2016-06-13   
2005        Boston Red Sox     Houston Astros  Fenway Park  2016-05-15   

      wind_speed  
1655        25.0  
2005        25.0

# 10. В какой игре получили максимальное количество очков?

столбец 'total_runs' показывает общее количество ранов для двух команд, соответственно необходимо взять максимальное значение этого столбца из датасета (первоначального, без фильтра по сезону)

In [40]:
print("Игры с максимальным количеством очков (ранов) (не важно какой сезон):")
data[data['total_runs'] == max(data['total_runs'])][['home_team', 'away_team', 'venue', 'date', 'total_runs']]

Игры с максимальным количеством очков (ранов) (не важно какой сезон):


home_team         away_team       venue        date  total_runs
1788  San Diego Padres  Seattle Mariners  Petco Park  2016-06-02          29

# 11. Какая игра содержала максимальное количество ошибок домашней команды?

столбец 'home_team_errors' показывает количество ошибок домашней команды, соответственно необходимо взять максимальное значение этого столбца из датасета (первоначального, без фильтра по сезону)

In [43]:
print("Матчи с максимальным количеством ошибок домашней команды (не важно какой сезон):")
data[data['home_team_errors'] == max(data['home_team_errors'])][['home_team', 'away_team', 'venue', 'date', 'home_team_errors']]

Матчи с максимальным количеством ошибок домашней команды (не важно какой сезон):


home_team             away_team        venue        date  \
1178  Milwaukee Brewers  Arizona Diamondbacks  Miller Park  2016-07-27   

      home_team_errors  
1178                 5

# 12. В какой игре было максимальное количество хитов? (исправила - было ранов)

столбцы 'away_team_hits' и 'home_team_hits' показывают количество хитов домашней и гостевой команды, нам необходимо сложить эти два столбца, чтобы получить общее количество хитов в игре ('total_hits') и далее необходимо взять максимальное значение этого столбца ('total_hits') из датасета (первоначального, без фильтра по сезону)

In [46]:
print("Игры с максимальным количеством хитов (не важно какой сезон):")
data['total_hits'] = data[['away_team_hits', 'home_team_hits']].sum(axis=1)
data[data['total_hits'] == max(data['total_hits'])][['home_team', 'away_team', 'venue', 'date','away_team_hits', 'home_team_hits', 'total_hits']]

Игры с максимальным количеством хитов (не важно какой сезон):


home_team      away_team        venue        date  away_team_hits  \
1413  Boston Red Sox  Texas Rangers  Fenway Park  2016-07-04              16   

      home_team_hits  total_hits  
1413              21          37

# 13. Выведите количество игр которая сыграла каждая команда в данном сезоне?

для решения этой задачи, мы можем создать два отдельных DataFrame, которые будут содержать информацию о количестве сыгранных игр для каждой команды, когда она была домашней и гостевой командой (из датасета с фильтром по регулярному сезону). Затем мы объединим эти два DataFrame и сгруппируем по названию команды, используя агрегирующую функцию sum.

In [49]:
number_of_games_home = data_rs[['home_team', 'home_team_win']].rename(columns={'home_team': 'team', 'home_team_win': 'result'}).groupby(['team']).count()
number_of_games_away = data_rs[['away_team', 'home_team_loss']].rename(columns={'away_team': 'team', 'home_team_loss': 'result'}).groupby(['team']).count()
print("Количество игр, которая сыграла каждая команда в данном сезоне (тип игрового сезона == 'regular season'):")
pd.concat([number_of_games_home, number_of_games_away]).groupby(['team']).sum('result').sort_values(by='result', ascending=False)

Количество игр, которая сыграла каждая команда в данном сезоне (тип игрового сезона == 'regular season'):


result
team                                 
Arizona Diamondbacks              162
Minnesota Twins                   162
Toronto Blue Jays                 162
Texas Rangers                     162
Tampa Bay Rays                    162
St. Louis Cardinals               162
Seattle Mariners                  162
San Francisco Giants              162
San Diego Padres                  162
Pittsburgh Pirates                162
Philadelphia Phillies             162
Oakland Athletics                 162
New York Yankees                  162
New York Mets                     162
Milwaukee Brewers                 162
Los Angeles Dodgers               162
Los Angeles Angels of Anaheim     162
Kansas City Royals                162
Houston Astros                    162
Colorado Rockies                  162
Cincinnati Reds                   162
Chicago White Sox                 162
Chicago Cubs                      162
Boston Red Sox                    162
Baltimore Orioles                 162
Washington Nationals              162
Atlanta Braves                    161
Miami Marlins                     161
Detroit Tigers                    161
Cleveland Indians                 161

# 14. Какая команда выиграла наибольшое количество матчей в сезоне? (Будьте внимательны с типом матча в сезоне - season).

для решения этой задачи, мы можем создать два отдельных DataFrame, которые будут содержать информацию о количестве сыгранных игр c победой для каждой команды, когда она была домашней и гостевой командой (из датасета с фильтром по регулярному сезону). Затем мы объединим эти два DataFrame и сгруппируем по названию команды, используя агрегирующую функцию sum. Используем сортировку по убыванию и выведем 1 строку.

In [52]:
number_of_games_home_win = data_rs[['home_team', 'home_team_win']].rename(columns={'home_team': 'team', 'home_team_win': 'result'}).groupby(['team']).sum()
number_of_games_away_win = data_rs[['away_team', 'home_team_loss']].rename(columns={'away_team': 'team', 'home_team_loss': 'result'}).groupby(['team']).sum()
print("Команда, которая выиграла наибольшое количество матчей в сезоне (тип игрового сезона == 'regular season'):")
pd.concat([number_of_games_home_win, number_of_games_away_win]).groupby(['team']).sum('result').sort_values(by='result', ascending=False).head(1)

Команда, которая выиграла наибольшое количество матчей в сезоне (тип игрового сезона == 'regular season'):


result
team                
Chicago Cubs     103

# 15. Какая команда выиграла наибольшее количество домашних матчей в сезоне?

столбец 'home_team_win' показывает 1 в случае победы домашней команды, соответственно необходимо сделать группировку по столбцу 'home_team' (название домашней команды), используя агрегирующую функцию sum (из датасета с фильтром по регулярному сезону)

In [55]:
print("Команда, которая выиграла наибольшее количество домашних матчей в сезоне (тип игрового сезона == 'regular season'):")
print(data_rs[['home_team', 'home_team_win']].groupby(['home_team']).sum().sort_values(by='home_team_win',ascending=False).head(1))

Команда, которая выиграла наибольшее количество домашних матчей в сезоне (тип игрового сезона == 'regular season'):
              home_team_win
home_team                  
Chicago Cubs             57


# 16. Какая команда выиграла наибольшее количество гостевых матчей в сезоне?

столбец 'home_team_loss' показывает 1 в случае победы гостевой команды, соответственно необходимо сделать группировку по столбцу 'away_team' (название гостевой команды), используя агрегирующую функцию sum (из датасета с фильтром по регулярному сезону)

In [58]:
print("Команда, которая выиграла наибольшее количество гостевых матчей в сезоне (тип игрового сезона == 'regular season'):")
print(data_rs[['away_team', 'home_team_loss']].groupby(['away_team']).sum().sort_values(by='home_team_loss',ascending=False).head(1))

Команда, которая выиграла наибольшее количество гостевых матчей в сезоне (тип игрового сезона == 'regular season'):
                     home_team_loss
away_team                          
St. Louis Cardinals              48


# 17. Какая команда проиграла наибольшее количество матчей в сезоне?

для решения этой задачи, мы можем создать два отдельных DataFrame, которые будут содержать информацию о количестве сыгранных игр c проигрышем для каждой команды, когда она была домашней и гостевой командой (из датасета с фильтром по регулярному сезону). Затем мы объединим эти два DataFrame и сгруппируем по названию команды, используя агрегирующую функцию sum. Используем сортировку по убыванию и выведем 1 строку.

In [61]:
number_of_games_home_loss = data_rs[['home_team', 'home_team_loss']].rename(columns={'home_team': 'team', 'home_team_loss': 'result'}).groupby(['team']).sum()
number_of_games_away_loss = data_rs[['away_team', 'home_team_win']].rename(columns={'away_team': 'team', 'home_team_win': 'result'}).groupby(['team']).sum()
print("Команда, которая проиграла наибольшое количество матчей в сезоне (тип игрового сезона == 'regular season'):")
pd.concat([number_of_games_home_loss, number_of_games_away_loss]).groupby(['team']).sum('result').sort_values(by='result', ascending=False).head(1)

Команда, которая проиграла наибольшое количество матчей в сезоне (тип игрового сезона == 'regular season'):


result
team                   
Minnesota Twins     103

# 18. Зависит ли выигрыш от количества посетителей матча? (https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.cov.html)

Вычислим ковариацию двух переменных 'attendance' - количество зрителей матча и 'home_team_win' - результат домашней команды (1 -  выигрыш) (предположим что в данной задаче речь идет о выигрыше домашней команды)

In [64]:
np.cov(data.attendance, data.home_team_win, ddof=0)

array([[9.73665253e+07, 2.24645590e+02],
       [2.24645590e+02, 2.49133543e-01]])

В нашем случае, положительное значение ковариации (224.645590) указывает на то, что, в целом, с увеличением количества зрителей наблюдается тенденция к увеличению вероятности победы домашней команды. Однако это значение само по себе не дает полной картины, так как ковариация зависит от единиц измерения переменных. Для более точного анализа рассмотрим коэффициент корреляции

In [66]:
data[['attendance', 'home_team_win']].corr()

attendance  home_team_win
attendance       1.000000       0.045612
home_team_win    0.045612       1.000000

In [67]:
st.pearsonr(data.attendance, data.home_team_win)

PearsonRResult(statistic=0.04561177364260062, pvalue=0.0235938997991939)

Получив значение корреляции (0.045612) и pvalue (0.0235938997991939) можно сделать вывод, что существует статистически значимая, но очень слабая положительная корреляция между количеством зрителей и победами домашней команды. Это может указывать на то, что, хотя и есть некоторая связь, она не является достаточно сильной, чтобы делать какие-либо уверенные выводы о влиянии количества зрителей на результат матча.

# 19. Правда ли что большинство проигрышных домашних матчей приходятся на Субботу и Воскресенье?

для того, чтобы решить эту задачу, нам необходимо посчитать число проигрышных матчей домашней команды для каждого дня недели

In [71]:
losses_by_day_of_the_week = data[data['home_team_win'] == 0][['day_of_week', 'home_team_win']].rename(columns={'home_team_win': 'loss'}).groupby(['day_of_week']).count()
losses_by_day_of_the_week.sort_values(by='loss', ascending=False)

loss
day_of_week      
Sunday        190
Tuesday       187
Friday        184
Saturday      175
Wednesday     167
Monday        131
Thursday      125

далее посчитаем количество матчей по дням недели

In [73]:
games_by_days_of_the_week = data[['day_of_week', 'home_team_win']].rename(columns={'home_team_win': 'games'}).groupby(['day_of_week']).count()
games_by_days_of_the_week.sort_values(by='games', ascending=False)

games
day_of_week       
Saturday       396
Friday         394
Sunday         392
Wednesday      379
Tuesday        376
Monday         278
Thursday       248

объединим эти две таблицы и посчитаем долю проигрышей для каждого дня недели

In [75]:
df = pd.merge(losses_by_day_of_the_week, games_by_days_of_the_week, left_index=True, right_index=True, how='outer')
df['percent_of_loss'] = df.apply(lambda x: x['loss']/x['games']*100, axis=1).round(2)
df.sort_values(by='percent_of_loss', ascending=False)

loss  games  percent_of_loss
day_of_week                              
Thursday      125    248            50.40
Tuesday       187    376            49.73
Sunday        190    392            48.47
Monday        131    278            47.12
Friday        184    394            46.70
Saturday      175    396            44.19
Wednesday     167    379            44.06

можно сделать вывод, что в пятницу, субботу и воскресенье проводится наибольшее количество игр, но при этом доля проигрышей у домашней команды в субботу и воскресенье не является наибольшей.

# 20. Правда ли что наибольшее количество ранов происходит в холодную погоду? (Холодной погодой считается погода ниже 0 градусов)

Дла решения этой задачи, добавим новый столбец ('cold_weather'), который даст нам значение True - Холодная погода  и False - Теплая погода (при этом обратим внимание что температура указана в фаренгейтах). Т.о. я буду считать, что погода холодная, если значение столбца 'temperature' менее 32 градусов по фаренгейту или 0 градусов по цельсию

In [79]:
data['cold_weather'] = data['temperature'] < 32
print(data['cold_weather'].sum())

1


из всего датасета (первоначального, без фильтра по сезону) всего одна игра проходила при температуре ниже 0 градусов по цельсию

In [81]:
data.groupby('cold_weather')['total_runs'].mean()

cold_weather
False     8.932169
True     12.000000
Name: total_runs, dtype: float64

не смотря на то, что среднее значение для игр при холодной погоде выше, чем при теплой погоде, выводы мы никакие сделать не можем, т.к. данных для анализа не достаточно